File: frontend/src/services/UserProgressService.ts
import axios from '../config/axiosConfig';
import { UserProgress } from '../types/UserProgress';

export class UserProgressService {
  private static readonly BASE_URL = '/api/progress';

  static async getRoadmapProgress(roadmapId: string): Promise<UserProgress> {
    const response = await axios.get(`${this.BASE_URL}/roadmap/${roadmapId}`);
    return response.data;
  }

  static async getUserProgress(): Promise<UserProgress[]> {
    const response = await axios.get(`${this.BASE_URL}/user`);
    return response.data;
  }

  static async createProgress(roadmapId: string): Promise<UserProgress> {
    const response = await axios.post(`${this.BASE_URL}/roadmap/${roadmapId}`);
    return response.data;
  }

  static async updateItemProgress(
    progressId: string,
    itemId: string,
    percentage: number
  ): Promise<UserProgress> {
    const response = await axios.put(
      `${this.BASE_URL}/${progressId}/item/${itemId}`,
      null,
      { params: { percentage } }
    );
    return response.data;
  }

  static async markItemCompleted(
    progressId: string,
    itemId: string
  ): Promise<UserProgress> {
    const response = await axios.put(
      `${this.BASE_URL}/${progressId}/item/${itemId}/complete`
    );
    return response.data;
  }

  static async updateTimeSpent(
    progressId: string,
    minutes: number
  ): Promise<UserProgress> {
    const response = await axios.put(
      `${this.BASE_URL}/${progressId}/time`,
      null,
      { params: { minutes } }
    );
    return response.data;
  }
}

File: backend/src/main/java/com/alphabrain/service/UserProgressService.java
package com.alphabrain.service;

import java.util.List;

import com.alphabrain.model.UserProgress;

public interface UserProgressService {
    
    // CRUD operations
    UserProgress createProgress(UserProgress progress);
    UserProgress updateProgress(String id, UserProgress progress);
    void deleteProgress(String id);
    UserProgress getProgress(String id);
    
    // User specific operations
    UserProgress getUserRoadmapProgress(String userId, String roadmapId);
    List<UserProgress> getUserProgress(String userId);
    List<UserProgress> getUserCompletedRoadmaps(String userId);
    List<UserProgress> getUserInProgressRoadmaps(String userId);
    List<UserProgress> getUserNotStartedRoadmaps(String userId);
    
    // Statistics
    long countUserCompletedRoadmaps(String userId);
    long countUserInProgressRoadmaps(String userId);
    double getAverageCompletionTime(String roadmapId);
    
    // Progress updates
    UserProgress updateItemProgress(String progressId, String itemId, int percentage);
    UserProgress markItemCompleted(String progressId, String itemId);
    UserProgress updateTimeSpent(String progressId, int additionalMinutes);
} 